<!doctype html><html lang="zh-tw">
<head>
    <meta http-equiv="refresh" content="0; URL=https://openhome.cc/Gossip/Python/'https://openhome.cc/zh-tw/python/exception/raise/'">
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="description" content="接續 try 陳述句 的內容。
      
except後若不接上任何例外型態，則表示捕捉所有例外，這包括了所有的系統例外，有時這並不是你想要的行為。例如，下面這個程式，無法透過Keyboard...">

<meta property="og:locale" content="zh_TW">
<meta property="og:title" content="再看 try、raise">
<meta property="og:type" content="article">
<meta property="og:url" content="https://openhome.cc/Gossip/Python/TryRaise.html">
<meta property="og:image" content="https://openhome.cc/Gossip/images/caterpillar_small.jpg">
<meta property="og:site_name" content="OPENHOME.CC">
<meta property="og:description" content="接續 try 陳述句 的內容。
      
except後若不接上任何例外型態，則表示捕捉所有例外，這包括了所有的系統例外，有時這並不是你想要的行為。例如，下面這個程式，無法透過Keyboard...">


    <title>再看 try、raise</title>

<link rel="stylesheet" href="../css/pure-0.6.0/pure-min.css">

    <!--[if lte IE 8]>
        <link rel="stylesheet" href="../css/layouts/side-menu-old-ie.css">
    <![endif]-->
    <!--[if gt IE 8]><!-->
        <link rel="stylesheet" href="../css/layouts/side-menu.css">
    <!--<![endif]-->
  

     <link rel="stylesheet" href="../css/caterpillar.css">
     <script async src="../google-code-prettify/run_prettify.js"></script>
<!-- 網頁層級廣告 --><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><script>(adsbygoogle =window.adsbygoogle || []).push({google_ad_client: "ca-pub-9750319131714390",enable_page_level_ads: true });</script></head>
<body>

<div id="layout">
    <!-- Menu toggle -->
    <a href="TryRaise.html#menu" id="menuLink" class="menu-link">
        <!-- Hamburger icon --><span></span>
        
    </a>
    <div id="menu">
        <div class="pure-menu">
            <a class="pure-menu-heading" href="https://openhome.cc/Gossip/Python/index.html">回 Python 目錄</a>
            <ul class="pure-menu-list">
                <br><div class="social" style="text-align: center;"><a href="http://twitter.com/caterpillar"><img title="Twitter" alt="Twitter" src="../images/twitter.png"></a> <a href="http://www.facebook.com/openhome.cc"><img title="Facebook" alt="Facebook" src="../images/facebook.png"></a> </div><br><div id="search box"><script>(function() {var cx = 'partner-pub-9750319131714390:3926766884';var gcse = document.createElement('script');gcse.type = 'text/javascript';gcse.async = true;gcse.src = (document.location.protocol == 'https:' ? 'https:' : 'http:') + '//www.google.com/cse/cse.js?cx=' + cx;var s = document.getElementsByTagName('script')[0];s.parentNode.insertBefore(gcse, s);})();</script><gcse:searchbox-only></gcse:searchbox-only></div><br><div class="ad" style="text-align: center;"><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
<!-- 2015 新版型 160 x 600 廣告 -->
<ins class="adsbygoogle"
     style="display:inline-block;width:160px;height:600px"
     data-ad-client="ca-pub-9750319131714390"
     data-ad-slot="3747048883"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script></div></li>
            </ul>
        </div>
    </div>

    <main id="main">
        <header class="header">
            <h1>再看 try、raise</h1>
        </header>

        <article class="content"><br><div class="ad-3" style="text-align: center;"><script async="" src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><!-- 2015 新版型回應式廣告 --><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-9750319131714390" data-ad-slot="7104125683" data-ad-format="auto"></ins><script>(adsbygoogle = window.adsbygoogle || []).push({});</script></div> 
            <span style="font-family: Courier New,Courier,monospace;">接續 </span><a href="TryStatement.html">try 陳述句</a> 的內容。<br>
      <span style="font-family: Courier New,Courier,monospace;"><br>
except</span>後若不接上任何例外型態，則表示捕捉所有例外，這包括了所有的系統例外，有時這並不是你想要的行為。例如，下面這個程式，無法透過<span style="font-family: Courier New,Courier,monospace;">KeyboardInterrupt</span>來中斷迴圈：<br>
      
      
      <div style="margin-left: 40px;"><span style="font-weight: bold; font-family: Courier New,Courier,monospace;">while True:</span><br>
      <span style="font-weight: bold; font-family: Courier New,Courier,monospace;">&nbsp;&nbsp;&nbsp; try:</span><br>
      <span style="font-weight: bold; font-family: Courier New,Courier,monospace;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; print('Run it....')</span><br>
      <span style="font-weight: bold; font-family: Courier New,Courier,monospace;">&nbsp;&nbsp;&nbsp; except:</span><br>
      <span style="font-weight: bold; font-family: Courier New,Courier,monospace;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; print('exception happened...')</span><br>
      </div>

      <br>

      由於except捕捉所有的例外，所以上例無法離開迴圈。你可以改為以下的方式：<br>
      
      
      <div style="margin-left: 40px;"><span style="font-weight: bold; font-family: Courier New,Courier,monospace;">while True:</span><br>
<span style="font-weight: bold; font-family: Courier New,Courier,monospace;">&nbsp;&nbsp;&nbsp; try:</span><br>
<span style="font-weight: bold; font-family: Courier New,Courier,monospace;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; print('Run it....')</span><br>
<span style="font-weight: bold; font-family: Courier New,Courier,monospace;">&nbsp;&nbsp;&nbsp; except Exception:</span><br>
<span style="font-weight: bold; font-family: Courier New,Courier,monospace;">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; print('exception happened...')</span></div>


      <br>

      在Python 3中，<span style="font-family: Courier New,Courier,monospace;">Exception是BaseException</span>的子類別，可以捕捉除了系統例外以外的所有例外。上例可以藉由<span style="font-family: Courier New,Courier,monospace;">KeyboardInterrupt中斷迴圈。</span><br>

      <br>

      在Python 3中，可以在<span style="font-family: Courier New,Courier,monospace;">except</span>捕捉到例外後，將例外物件指定給變數。例如：<br>
      <br>
      
      
      <div class="cmd"><span><span style="font-weight: bold;">&gt;&gt;&gt; try:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; raise IndexError('11')<br>
... except IndexError as e:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(type(e), str(e))<br>
...<br>
</span>&lt;class 'IndexError'&gt; 11<span style="font-weight: bold;"><br>
&gt;&gt;&gt;</span><span style="font-weight: bold;"></span><span style="font-weight: bold;"></span></span><span><br>




            </span></div>





      <br>

      <span style="font-family: Courier New,Courier,monospace;">在更進階的例外追蹤需求中，可以使用<span style="font-weight: bold;">sys.exc_info()</span>方法取得一個</span>Tuple<span style="font-family: Courier New,Courier,monospace;">物件，該Tuple物件中包括了例外的類型、例外訊息以及<span style="font-weight: bold;">traceback</span>物件：</span><br>
      
      <div class="cmd"><span><span style="font-weight: bold;">&gt;&gt;&gt; try:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; raise 'error'<br>
... except:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; a, b, c = sys.exc_info()<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(a)<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(b)<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(c)<br>
...<br>
</span>&lt;class 'TypeError'&gt;<br>
exceptions must derive from BaseException<br>
&lt;traceback object at 0x01D4FB20&gt;<span style="font-weight: bold;"><br>
&gt;&gt;&gt;</span><span style="font-weight: bold;"></span><span style="font-weight: bold;"></span></span><span><br>




            </span></div>





      <br>

      <span style="font-family: Courier New,Courier,monospace;">trackback物件代表了呼叫堆疊中每一個層次的追蹤，可以使用<span style="font-weight: bold;">tb_next</span>取得更深一層的呼叫堆疊。例如：<br>
      </span><span style="font-family: Courier New,Courier,monospace;"></span>
      
      <pre class="prettyprint"><code>import sys<br>def test():<br>    raise EOFError<br><br>try:<br>    test()<br>except:<br>    type, message, traceback = sys.exc_info()<br>    while traceback:<br>        print('..........')<br>        print(type)<br>        print(message)<br>        print('function or module？', traceback.tb_frame.f_code.co_name)<br>        print('file？', traceback.tb_frame.f_code.co_filename)<br>        traceback = traceback.tb_next<br></code></pre>

      <br>
      <span style="font-family: Courier New,Courier,monospace;"><span style="font-weight: bold;">tb_frame</span>代表了</span>該層追蹤<span style="font-family: Courier New,Courier,monospace;">的所有物件資訊，<span style="font-weight: bold;">f_code</span>可以取得該層的程式碼資訊，例如<span style="font-weight: bold;">co_name</span>可取得函式或模組名稱，而<span style="font-weight: bold;">co_filename</span>則表示該程式碼所在的檔案。上例的執行範例如下：</span><br>
      
      <div class="cmd"><span>..........<br>
&lt;class 'EOFError'&gt;<br>
            <br>
function or module？ &lt;module&gt;<br>
file？ demo.py<br>
..........<br>
&lt;class 'EOFError'&gt;<br>
            <br>
function or module？ test<br>
file？ demo.py<br>
<span style="font-weight: bold;"></span></span></div>






      <br>
      再來看看raise的使用，正如在 <span style="font-family: Courier New,Courier,monospace;"></span><a href="TryStatement.html">try 陳述句</a> 中看到的，你可以在<span style="font-family: Courier New,Courier,monospace;">raise</span>後接上字串或例外類別名稱，現在已不鼓勵<span style="font-family: Courier New,Courier,monospace;">raise字串實例。實際上，raise</span>之後可以接上例外類別名稱、例外實例或不接上任何東西。<br>
      <br>
當你在<span style="font-family: Courier New,Courier,monospace;">raise</span>後接上例外類別時，實際上會以該類別建立實例再丟出，也就是下面兩行程式碼的作用是相同的：<br>
      
      <div style="margin-left: 40px;"><span style="font-weight: bold; font-family: Courier New,Courier,monospace;">raise EOFError</span><br>
      <span style="font-weight: bold; font-family: Courier New,Courier,monospace;">raise EOFError()</span><br>
      </div>

      <br>
      如果在except中使用raise而不接上任何物件，則表示將except比對到的例外實例再度丟出。例如：<br>
      
      
      <div class="cmd"><span><span style="font-weight: bold;">&gt;&gt;&gt; try:</span><br>
            <span style="font-weight: bold;">...&nbsp;&nbsp;&nbsp;&nbsp; raise EOFError</span><br>
            <span style="font-weight: bold;">... except EOFError:</span><br>
            <span style="font-weight: bold;">...&nbsp;&nbsp;&nbsp;&nbsp; print('got it')</span><br>
            <span style="font-weight: bold;">...&nbsp;&nbsp;&nbsp;&nbsp; raise</span><br>
            <span style="font-weight: bold;">...</span><br>
got it<br>
Traceback (most recent call last):<br>
&nbsp; File "&lt;stdin&gt;", line 2, in &lt;module&gt;<br>
EOFError<br>
&gt;&gt;&gt;<br>
<span style="font-weight: bold;"></span></span></div>







      <br>
      如果必要的話，你還可以在<span style="font-family: Courier New,Courier,monospace;">except中raise例外時，附上except</span>所比對到的例外實例。例如。<br>
      <div class="cmd"><span><span style="font-weight: bold;">&gt;&gt;&gt; try:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; raise EOFError<br>
... except EOFError as e:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print('got it')<br>
...&nbsp;&nbsp;&nbsp;&nbsp; raise IndexError from e<br>
...<br>
</span>got it<br>
Traceback (most recent call last):<br>
&nbsp; File "&lt;stdin&gt;", line 2, in &lt;module&gt;<br>
EOFError<br>
            <br>
The above exception was the direct cause of the following exception:<br>
            <br>
Traceback (most recent call last):<br>
&nbsp; File "&lt;stdin&gt;", line 5, in &lt;module&gt;<br>
IndexError<span style="font-weight: bold;"><br>
&gt;&gt;&gt;</span><br>
<span style="font-weight: bold;"></span></span></div>








      <br>
      <span style="font-weight: bold;">raise from</span>語法，會將from後接上的例外實例，設定給被raise的例外實例之<span style="font-weight: bold;">__cause__</span>。例如：<br>

      <div class="cmd"><span><span style="font-weight: bold;">&gt;&gt;&gt; try:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; try:<br>
...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; raise EOFError('XD')<br>
...&nbsp;&nbsp;&nbsp;&nbsp; except EOFError as e:<br>
...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; print(e.args)<br>
...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; raise IndexError('Orz') from e<br>
... except IndexError as e:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(e.args)<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(e.__cause__.args)<br>
...<br>
</span>('XD',)<br>
('Orz',)<br>
('XD',)<span style="font-weight: bold;"><br>
&gt;&gt;&gt;</span><span style="font-weight: bold;"></span><br>
<span style="font-weight: bold;"></span></span></div>









      <br>
      實際上，如果你在except中raise某個例外，則原except所比對到的例外，無論有無使用raise from，都會自動設定給<span style="font-weight: bold;">__context__</span>。例如：<br>
      <div class="cmd"><span><span style="font-weight: bold;">&gt;&gt;&gt; try:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; try:<br>
...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; raise EOFError('XD')<br>
...&nbsp;&nbsp;&nbsp;&nbsp; except EOFError as e:<br>
...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; print(e.args)<br>
...&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; raise IndexError('Orz') from e<br>
... except IndexError as e:<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(e.args)<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(e.__cause__.args)<br>
...&nbsp;&nbsp;&nbsp;&nbsp; print(e.__context__.args)<br>
...</span><br>
('XD',)<br>
('Orz',)<br>
('XD',)<br>
('XD',)<span style="font-weight: bold;"><br>
&gt;&gt;&gt;</span><span style="font-weight: bold;"></span><span style="font-weight: bold;"></span><br>
<span style="font-weight: bold;"></span></span></div>










      <br>




<br><br><div class="ad336-280" style="text-align: center;"><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><!-- 2015 新版型廣告 336 x 280 --><ins class="adsbygoogle" style="display:inline-block;width:336px;height:280px" data-ad-client="ca-pub-9750319131714390" data-ad-slot="9976409681"></ins><script>(adsbygoogle = window.adsbygoogle || []).push({});</script></div><br><div class="recommend" style="text-align: center;"><hr><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><!-- 自動大小回應相符內容 --><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-9750319131714390" data-ad-slot="4953478487" data-ad-format="autorelaxed"></ins><script>(adsbygoogle = window.adsbygoogle || []).push({});</script></div></article></main></div></body>
<script src="../js/ui.js"></script>
</html>
<div class="analytics"><script async src="https://www.googletagmanager.com/gtag/js?id=G-QVQQYFSC8J"></script><script>window.dataLayer = window.dataLayer || [];function gtag(){dataLayer.push(arguments);}gtag('js', new Date());gtag('config', 'G-QVQQYFSC8J');</script></div>
